pkg/controllermanager/controller/shoot/shoot_retry_control.go

```diff
@@ -0,0 +1,140 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func (c *Controller) shootRetryAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.shootRetryQueue.Add(key)
+}
+
+func (c *Controller) shootRetryUpdate(oldObj, newObj interface{}) {
+	var (
+		oldShoot = oldObj.(*gardencorev1beta1.Shoot)
+		newShoot = newObj.(*gardencorev1beta1.Shoot)
+	)
+
+	if shootFailedDueToRateLimits(newShoot) && !isShootFailed(oldShoot) {
+		key, err := cache.MetaNamespaceKeyFunc(newObj)
+		if err != nil {
+			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", newObj, err)
+			return
+		}
+		c.shootRetryQueue.Add(key)
+	}
+}
+
+// NewShootRetryReconciler creates a new instance of a reconciler which retries certain failed Shoots.
+func NewShootRetryReconciler(l logrus.FieldLogger, gardenClient client.Client, config *config.ShootRetryControllerConfiguration) reconcile.Reconciler {
+	return &shootRetryReconciler{
+		logger:       l,
+		gardenClient: gardenClient,
+		config:       config,
+	}
+}
+
+type shootRetryReconciler struct {
+	logger       logrus.FieldLogger
+	gardenClient client.Client
+	config       *config.ShootRetryControllerConfiguration
+}
+
+func (r *shootRetryReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := r.gardenClient.Get(ctx, request.NamespacedName, shoot); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Infof("Object %q is gone, stop reconciling: %v", request.Name, err)
+			return reconcile.Result{}, nil
+		}
+		r.logger.Infof("Unable to retrieve object %q from store: %v", request.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	key := fmt.Sprintf("%s/%s", shoot.Namespace, shoot.Name)
+	shootLogger := r.logger.WithField("shoot", key)
+
+	if !shootFailedDueToRateLimits(shoot) {
+		return reconcile.Result{}, nil
+	}
+
+	mustRetry, requeueAfter := mustRetryNow(shoot, *r.config.RetryPeriod)
+	if !mustRetry {
+		shootLogger.Infof("[SHOOT RETRY] Scheduled retry in %s", requeueAfter.Round(time.Minute))
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
+	}
+
+	shootLogger.Info("[SHOOT RETRY] Retrying a failed Shoot")
+
+	shootCopy := shoot.DeepCopy()
+	metav1.SetMetaDataAnnotation(&shoot.ObjectMeta, v1beta1constants.GardenerOperation, v1beta1constants.ShootOperationRetry)
+	if err := r.gardenClient.Patch(ctx, shoot, client.MergeFrom(shootCopy)); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	shootLogger.Info("[SHOOT RETRY] Shoot was successfully retried")
+
+	return reconcile.Result{}, nil
+}
+
+func shootFailedDueToRateLimits(shoot *gardencorev1beta1.Shoot) bool {
+	return isShootFailed(shoot) && gardencorev1beta1helper.HasErrorCode(shoot.Status.LastErrors, gardencorev1beta1.ErrorInfraRateLimitsExceeded)
+}
+
+func isShootFailed(shoot *gardencorev1beta1.Shoot) bool {
```

Comment:
Could this be replaced with https://github.com/gardener/gardener/blob/a4a55c852b1486a4aaa3372771ab7b365737d3a3/pkg/utils/gardener/shoot.go#L52-L57?

Reply:
The reason I introduced a new func is that the existing one check also the `.status.gardener.version` field with `shoot.Status.Gardener.Version == version.Get().GitVersion`. This version check is valid when it is performed by gardenlet. On the other side GCM can run another version (custom build) and I think other components (except gardenlet) should not check the `.status.gardener.version` when checking if a Shoot is failed or not (currently the extension library also uses a func that does not check the `.status.gardener.version`). That's why I didn't reuse the existing `IsShootFailed` func.